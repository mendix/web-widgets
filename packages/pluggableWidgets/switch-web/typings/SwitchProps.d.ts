/**
 * This file was generated from Switch.xml
 * WARNING: All changes made to this file will be overwritten
 * @author Mendix Widgets Framework Team
 */
import { ActionValue, EditableValue } from "mendix";

export interface SwitchContainerProps {
    name: string;
    tabIndex?: number;
    id: string;
    booleanAttribute: EditableValue<boolean>;
    action?: ActionValue;
}

export interface SwitchPreviewProps {
    readOnly: boolean;
    booleanAttribute: string;
    action: {} | null;
}
