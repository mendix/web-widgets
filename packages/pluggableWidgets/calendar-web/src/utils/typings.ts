import { ObjectItem } from "mendix";
import { CalendarProps as ReactCalendarProps } from "react-big-calendar";
import { withDragAndDropProps } from "react-big-calendar/lib/addons/dragAndDrop";

export interface CalendarEvent {
    title: string;
    start: Date;
    end: Date;
    allDay: boolean;
    color?: string;
    item: ObjectItem;
}

export type EventDropOrResize = {
    event: CalendarEvent;
    start: Date;
    end: Date;
};

export interface DragAndDropCalendarProps<TEvent extends object = Event, TResource extends object = object>
    extends ReactCalendarProps<TEvent, TResource>,
        withDragAndDropProps<TEvent, TResource> {}
