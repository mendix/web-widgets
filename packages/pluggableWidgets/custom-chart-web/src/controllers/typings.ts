import { ActionValue, EditableValue } from "mendix";

export interface ControllerProps {
    configurationOptions: string;
    dataAttribute?: EditableValue<string>;
    dataStatic: string;
    eventDataAttribute?: EditableValue<string>;
    layoutAttribute?: EditableValue<string>;
    layoutStatic: string;
    onClick?: ActionValue;
    sampleData: string;
    sampleLayout: string;
}
