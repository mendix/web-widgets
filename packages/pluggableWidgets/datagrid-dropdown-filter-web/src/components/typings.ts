import { RefFilterStore } from "@mendix/widget-plugin-dropdown-filter/stores/RefFilterStore";
import { StaticSelectFilterStore } from "@mendix/widget-plugin-dropdown-filter/stores/StaticSelectFilterStore";

export interface EnumFilterAPI {
    filterStore: StaticSelectFilterStore;
    parentChannelName?: string;
}

export interface RefFilterAPI {
    filterStore: RefFilterStore;
    parentChannelName?: string;
}
