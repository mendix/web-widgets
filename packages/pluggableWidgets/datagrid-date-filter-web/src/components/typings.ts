import { Date_InputFilterInterface } from "@mendix/widget-plugin-filtering/typings/InputFilterInterface";

export interface DateFilterProps {
    filterStore: Date_InputFilterInterface;
    parentChannelName?: string;
}
