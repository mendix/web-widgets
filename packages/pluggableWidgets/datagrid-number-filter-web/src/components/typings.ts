import { ChangeEventHandler, CSSProperties } from "react";
import { FilterType } from "../../typings/FilterType";
import { Big } from "big.js";

interface FilterProps {
    adjustable: boolean;
    defaultFilterType: FilterType;
    className?: string;
    id?: string;
    placeholder?: string;
    screenReaderButtonCaption?: string;
    screenReaderInputCaption?: string;
    tabIndex?: number;
    styles?: CSSProperties;
}

export interface FilterInputProps extends FilterProps {
    onFilterTypeClick: (type: FilterType) => void;
    onInputChange: ChangeEventHandler<HTMLInputElement>;
    inputValue: string;
    inputRef?: React.ClassAttributes<HTMLInputElement>["ref"];
    inputDisabled?: boolean;
}

export interface FilterComponentProps extends FilterProps {
    inputChangeDelay: number;
    value?: Big;
    updateFilters?: (value: Big | undefined, type: FilterType) => void;
    name: string;
    datagridChannelName?: string;
}
