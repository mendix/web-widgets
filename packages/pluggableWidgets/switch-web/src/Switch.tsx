import { createElement, KeyboardEvent, MouseEvent, useCallback } from "react";
import { executeAction } from "@mendix/widget-plugin-platform/framework/execute-action";
import { isAvailable } from "@mendix/widget-plugin-platform/framework/is-available";
import SwitchComponent from "./components/Switch";

import { SwitchContainerProps } from "../typings/SwitchProps";
import "./ui/switch-main.scss";

export function Switch(props: SwitchContainerProps) {
    const isChecked = isAvailable(props.booleanAttribute);
    const editable = !props.booleanAttribute.readOnly;

    const toggle = useCallback(() => {
        props.booleanAttribute.setValue(!props.booleanAttribute.value);
        executeAction(props.action);
    }, [props.action, props.booleanAttribute]);

    const onClick = useCallback(
        (event: MouseEvent<HTMLDivElement>) => {
            event.preventDefault();
            if (editable) {
                toggle();
            }
        },
        [editable, toggle]
    );
    const onKeyDown = useCallback(
        (event: KeyboardEvent<HTMLDivElement>) => {
            event.preventDefault();
            if (editable && event.key === " ") {
                toggle();
            }
        },
        [editable, toggle]
    );

    return (
        <SwitchComponent
            onClick={onClick}
            onKeyDown={onKeyDown}
            isChecked={isChecked}
            editable={editable}
            validation={props.booleanAttribute.validation}
            id={props.id}
            tabIndex={props.tabIndex}
        />
    );
}
