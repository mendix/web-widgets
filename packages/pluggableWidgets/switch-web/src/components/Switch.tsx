import { MouseEvent, KeyboardEvent, createElement } from "react";
import classNames from "classnames";
import { Alert } from "@mendix/widget-plugin-component-kit/Alert";
import { SwitchContainerProps } from "../../typings/SwitchProps";

export interface SwitchProps extends Pick<SwitchContainerProps, "id" | "tabIndex"> {
    onClick?: (event: MouseEvent<HTMLDivElement>) => void;
    onKeyDown?: (event: KeyboardEvent<HTMLDivElement>) => void;
    isChecked: boolean;
    editable: boolean;
    validation: string | undefined;
}

export default function Switch(props: SwitchProps) {
    return (
        <div className="widget-switch">
            <div
                id={props.id}
                className={classNames("widget-switch-btn-wrapper", "widget-switch-btn-wrapper-default", {
                    checked: props.isChecked,
                    disabled: !props.editable,
                    "un-checked": !props.isChecked
                })}
                onClick={props.onClick}
                onKeyDown={props.onKeyDown}
                tabIndex={props.tabIndex ?? 0}
                role="switch"
                aria-checked={props.isChecked}
                aria-labelledby={`${props.id}-label`}
                aria-readonly={!props.editable}
            >
                <div
                    className={classNames("widget-switch-btn", {
                        left: !props.isChecked,
                        right: props.isChecked
                    })}
                />
            </div>
            <Alert bootstrapStyle={"danger"}>{props.validation}</Alert>
        </div>
    );
}
