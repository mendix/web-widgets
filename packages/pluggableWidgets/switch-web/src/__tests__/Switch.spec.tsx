import { Alert, AlertProps } from "@mendix/widget-plugin-component-kit/Alert";
import { actionValue, EditableValueBuilder } from "@mendix/widget-plugin-test-utils";
import { mount, ReactWrapper } from "enzyme";
import { createElement } from "react";

import SwitchComponent, { SwitchProps } from "../components/Switch";
import { SwitchContainerProps } from "../../typings/SwitchProps";
import { Switch } from "../Switch";

describe("Switch", () => {
    let switchWrapper: ReactWrapper<SwitchContainerProps, any>;
    let switchComponent: ReactWrapper<SwitchProps, any>;
    let switchButtonWrapper: ReactWrapper<any, any>;
    let switchButton: ReactWrapper<any, any>;
    let alert: ReactWrapper<AlertProps, any>;
    const createAndFindElements = (props: SwitchContainerProps): void => {
        switchWrapper = mount(<Switch {...props} />);
        switchComponent = switchWrapper.find(SwitchComponent);
        switchButtonWrapper = switchComponent.find(".widget-switch-btn-wrapper");
        switchButton = switchComponent.find(".widget-switch-btn");
        alert = switchComponent.find(Alert);
    };
    const createProps = (props?: Partial<SwitchContainerProps>): SwitchContainerProps => {
        const defaultProps: SwitchContainerProps = {
            name: "switch",
            tabIndex: 0,
            id: "com.mendix.widgets.custom.switch1",
            booleanAttribute: new EditableValueBuilder<boolean>().withValue(false).build(),
            action: undefined
        };

        return { ...defaultProps, ...props };
    };

    it("with editable value renders the structure correctly", () => {
        createAndFindElements(
            createProps({
                booleanAttribute: new EditableValueBuilder<boolean>().withValue(false).build()
            })
        );

        expect(switchWrapper).toMatchSnapshot();
    });

    it("with readonly value renders the structure correctly", () => {
        createAndFindElements(
            createProps({
                booleanAttribute: new EditableValueBuilder<boolean>().isReadOnly().withValue(false).build()
            })
        );

        expect(switchWrapper).toMatchSnapshot();
    });

    it("without validation message renders correctly", () => {
        createAndFindElements(
            createProps({
                booleanAttribute: new EditableValueBuilder<boolean>().withValue(false).build()
            })
        );

        expect(alert).toMatchSnapshot();
    });

    it("with validation message renders correctly", () => {
        createAndFindElements(
            createProps({
                booleanAttribute: new EditableValueBuilder<boolean>().withValidation("error").withValue(false).build()
            })
        );

        expect(alert).toMatchSnapshot();
    });

    it("when value is false renders with correct attributes", () => {
        createAndFindElements(createProps());

        expect(switchButtonWrapper.hasClass("un-checked")).toBe(true);
        expect(switchButtonWrapper.hasClass("checked")).toBe(false);
        expect(switchButtonWrapper.props()["aria-checked"]).toBe(false);
        expect(switchButton.hasClass("left")).toBe(true);
        expect(switchButton.hasClass("right")).toBe(false);
    });

    it("when value is true renders with correct attributes", () => {
        createAndFindElements(
            createProps({ booleanAttribute: new EditableValueBuilder<boolean>().withValue(true).build() })
        );

        expect(switchButtonWrapper.hasClass("un-checked")).toBe(false);
        expect(switchButtonWrapper.hasClass("checked")).toBe(true);
        expect(switchButtonWrapper.props()["aria-checked"]).toBe(true);
        expect(switchButton.hasClass("left")).toBe(false);
        expect(switchButton.hasClass("right")).toBe(true);
    });

    it("with tabIndex passed renders correctly", () => {
        createAndFindElements(createProps({ tabIndex: 1 }));

        expect(switchButtonWrapper.props().tabIndex).toEqual(1);
    });

    it("without tabIndex passed renders correctly", () => {
        createAndFindElements(createProps({ tabIndex: undefined }));

        expect(switchButtonWrapper.props().tabIndex).toEqual(0);
    });

    describe("when editable", () => {
        it("renders elements with correct attributes", () => {
            createAndFindElements(createProps());

            expect(switchButtonWrapper.hasClass("disabled")).toBe(false);
            expect(switchButtonWrapper.props()["aria-readonly"]).toBe(false);
        });

        it("invokes preventDefault onClick", () => {
            const props = createProps({ action: actionValue() });
            createAndFindElements(props);
            const eventMock = { preventDefault: jest.fn() };

            switchButtonWrapper.simulate("click", eventMock);

            expect(eventMock.preventDefault).toHaveBeenCalled();
        });

        it("invokes preventDefault on space keydown", () => {
            const props = createProps({ action: actionValue() });
            createAndFindElements(props);
            const eventMock = { preventDefault: jest.fn(), key: " " };

            switchButtonWrapper.simulate("keydown", eventMock);

            expect(eventMock.preventDefault).toHaveBeenCalled();
        });

        it("invokes action on click", () => {
            const props = createProps({ action: actionValue() });
            createAndFindElements(props);

            switchButtonWrapper.simulate("click");

            expect(props.action?.execute).toHaveBeenCalled();
        });

        it("invokes action on space keydown", () => {
            const props = createProps({ action: actionValue() });
            createAndFindElements(props);

            switchButtonWrapper.simulate("keydown", { key: " " });

            expect(props.action?.execute).toHaveBeenCalled();
        });

        it("shouldn't invoke action on keydown of any key but space", () => {
            const props = createProps({ action: actionValue() });
            createAndFindElements(props);

            switchButtonWrapper.simulate("keydown", { key: "enter" });

            expect(props.action?.execute).not.toHaveBeenCalled();
        });

        describe("when value is available", () => {
            it("toggles the attributes value onClick", () => {
                const props = createProps({
                    booleanAttribute: new EditableValueBuilder<boolean>().withValue(false).build()
                });
                createAndFindElements(props);

                switchButton.simulate("click");

                expect(props.booleanAttribute.setValue).toHaveBeenCalled();
                expect(props.booleanAttribute.value).toEqual(true);

                switchButton.simulate("click");

                expect(props.booleanAttribute.setValue).toHaveBeenCalled();
                expect(props.booleanAttribute.value).toEqual(false);
            });

            it("toggles the attribute value on space keydown", () => {
                const props = createProps({
                    booleanAttribute: new EditableValueBuilder<boolean>().withValue(false).build()
                });
                createAndFindElements(props);

                switchButton.simulate("keydown", { key: " " });

                expect(props.booleanAttribute.setValue).toHaveBeenCalled();
                expect(props.booleanAttribute.value).toEqual(true);

                switchButton.simulate("keydown", { key: " " });

                expect(props.booleanAttribute.setValue).toHaveBeenCalled();
                expect(props.booleanAttribute.value).toEqual(false);
            });

            it("shouldn't toggle the attribute on keydown of any key but space", () => {
                const props = createProps({
                    booleanAttribute: new EditableValueBuilder<boolean>().withValue(false).build()
                });
                createAndFindElements(props);

                switchButton.simulate("keydown", { key: "enter" });

                expect(props.booleanAttribute.setValue).not.toHaveBeenCalled();
            });
        });
    });

    describe("when readonly", () => {
        it("renders elements with correct attributes", () => {
            createAndFindElements(
                createProps({ booleanAttribute: new EditableValueBuilder<boolean>().isReadOnly().build() })
            );

            expect(switchButtonWrapper.hasClass("disabled")).toBe(true);
            expect(switchButtonWrapper.props()["aria-readonly"]).toBe(true);
        });

        it("shouldn't invoke action", () => {
            const props = createProps({
                booleanAttribute: new EditableValueBuilder<boolean>().isReadOnly().build(),
                action: actionValue()
            });
            createAndFindElements(props);

            switchButton.simulate("click");

            expect(props.action?.execute).not.toHaveBeenCalled();
        });

        it("shouldn't change the attributes value", () => {
            const props = createProps({
                booleanAttribute: new EditableValueBuilder<boolean>().isReadOnly().build()
            });
            createAndFindElements(props);

            switchButton.simulate("click");

            expect(props.booleanAttribute.setValue).not.toHaveBeenCalled();
        });

        it("invokes preventDefault onClick", () => {
            const props = createProps({ action: actionValue() });
            createAndFindElements(props);
            const eventMock = { preventDefault: jest.fn() };

            switchButtonWrapper.simulate("click", eventMock);

            expect(eventMock.preventDefault).toHaveBeenCalled();
        });

        it("invokes preventDefault on keydown space", () => {
            const props = createProps({ action: actionValue() });
            createAndFindElements(props);
            const eventMock = { preventDefault: jest.fn(), key: " " };

            switchButtonWrapper.simulate("keydown", eventMock);

            expect(eventMock.preventDefault).toHaveBeenCalled();
        });
    });
});
