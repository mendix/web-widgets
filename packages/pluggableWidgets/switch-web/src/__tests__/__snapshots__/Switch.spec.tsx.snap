// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Switch with editable value renders the structure correctly 1`] = `
<div>
  <div
    class="widget-switch"
  >
    <input
      aria-hidden="true"
      id="com.mendix.widgets.custom.switch1"
      style="opacity: 0; position: absolute; width: 1px; height: 1px;"
      tabindex="-1"
      type="checkbox"
    />
    <div
      aria-checked="false"
      aria-labelledby="com.mendix.widgets.custom.switch1-label"
      aria-readonly="false"
      class="widget-switch-btn-wrapper widget-switch-btn-wrapper-default un-checked"
      role="switch"
      tabindex="0"
    >
      <div
        class="widget-switch-btn left"
      />
    </div>
  </div>
</div>
`;

exports[`Switch with readonly value renders the structure correctly 1`] = `
<div>
  <div
    class="widget-switch"
  >
    <input
      aria-hidden="true"
      disabled=""
      id="com.mendix.widgets.custom.switch1"
      style="opacity: 0; position: absolute; width: 1px; height: 1px;"
      tabindex="-1"
      type="checkbox"
    />
    <div
      aria-checked="false"
      aria-labelledby="com.mendix.widgets.custom.switch1-label"
      aria-readonly="true"
      class="widget-switch-btn-wrapper widget-switch-btn-wrapper-default disabled un-checked"
      role="switch"
      tabindex="0"
    >
      <div
        class="widget-switch-btn left"
      />
    </div>
  </div>
</div>
`;
